src/os_posix/os_mtx_cond.c

```diff
@@ -103,27 +103,30 @@ void
 __wt_cond_signal(WT_SESSION_IMPL *session, WT_CONDVAR *cond)
 {
 	WT_DECL_RET;
-	bool locked;
-
-	locked = false;
 
 	__wt_verbose(session, WT_VERB_MUTEX, "signal %s", cond->name);
 
-	/* Fast path if already signalled. */
-	if (cond->waiters == -1)
-		return;
+	/*
+	 * Our callers are often setting flags to cause a thread to exit. Add
+	 * a barrier to ensure the flags are seen by the threads.
+	 */
+	WT_WRITE_BARRIER();
 
-	if (cond->waiters > 0 || !__wt_atomic_casi32(&cond->waiters, 0, -1)) {
-		WT_ERR(pthread_mutex_lock(&cond->mtx));
-		locked = true;
-		WT_ERR(pthread_cond_broadcast(&cond->cond));
-	}
+	/*
+	 * Fast path if we are in (or can enter), a state where the next waiter
+	 * will return immediately as already signaled.
+	 */
+	if (cond->waiters == -1 ||
+	    (cond->waiters == 0 && __wt_atomic_casi32(&cond->waiters, 0, -1)))
```

Comment:
The hang happens rarely enough we won't know until the branch has been running awhile. I think this branch fixes a bug (flags to cause the server thread to exit not getting flushed), and makes things better/cleaner, but I'm not sure it fixes the hang.

I don't think that reading `cond->waiters` multiple times is the problem: if the first read matches -1 then we must have read `waiters` before any threads waiting on the mutex, and the next thread attempting a wait will return immediately; if the first read isn't -1 and the second read matches 0, then we atomically swap `waiters` to -1: any thread trying to wait atomically increments `waiters` and so either that thread wins and our swap must fail and we'll proceed to wake the thread up, or we win and our swap to -1 followed by that thread's increment will force that thread to return immediately.

I *think*. :)

I'd be fine with turning that code into a single read, but I don't see a bug there.